lib/iris/fileformats/grib/_message.py

```diff
@@ -110,6 +108,8 @@ def data(self):
                 raise TranslationError(msg)
             if template == 90:
                 shape = (grid_section['Ny'], grid_section['Nx'])
+            elif template == 40 and reduced:
+                shape = (grid_section['numberOfDataPoints'],)
```

Comment:
I working on the assumption that the `_message` module should be independent of the `_load_convert` module. It's just a low-level interface to GRIB messages and knows nothing of how to convert to cubes. (At some point we might want to make it a public API - perhaps in a different package. Note how it has almost no dependency on other Iris modules.)

In which case, `_message` shouldn't be relying on `_load_convert` to do all the sanity checking. If `reduced` is set then it seems we should either deal with it or raise an error.